google/cloud/storage/internal/grpc/object_request_parser.cc

```diff
@@ -388,9 +388,15 @@ StatusOr<google::storage::v2::ReadObjectRequest> ToProto(
   }
   if (request.HasOption<storage::ReadFromOffset>()) {
```

Comment:
I get the math, but it's confusing to read. Consider only setting `read_offset`, `read_limit` once for any configuration of options. e.g.:

```cc
  auto const offset = request.HasOption<storage::ReadFromOffset>()
                          ? request.GetOption<storage::ReadFromOffset>().value()
                          : 0;
  if (request.HasOption<storage::ReadRange>()) {
    auto const range = request.GetOption<storage::ReadRange>().value();
    auto const start = (std::max)(range.begin, offset);
    auto const limit = range.end - start;
    if (limit <= 0) return internal::InvalidArgumentError("blah blah");

    r.set_read_offset(start);
    r.set_read_limit(limit);
  } else if (offset != 0) {
    r.set_read_offset(offset);
  }
```

Reply:
It would not work if negative ranges are a thing. I assume `range.begin >= 0`.

Reply:
Negative ranges are not a thing, that is not something you can send to the service.  I am not overly concerned about detecting or handling invalid uses of these ranges.

I like your code (modulo minor stuff) better. PTAL.